dart/collision/CollisionGroup.h

```diff
@@ -0,0 +1,210 @@
+
+  /// Unregister all the BodyNode's ShapeFrames having CollisionAddon from this
+  /// CollisionGroup
+  void unregisterShapeFramesOf(const dynamics::BodyNode* bodyNode);
+
+  /// Unregister all the Skeleton's ShapeFrames having CollisionAddon from this
+  /// CollisionGroup
+  void unregisterShapeFramesOf(const dynamics::Skeleton* skeleton);
+
+  /// Unregister all the ShapeFrames in this CollisionGroup
+  void unregisterAllShapeFrames();
+
+  /// Return true if this CollisionGroup contains shapeFrame
+  bool hasShapeFrame(const dynamics::ShapeFrame* shapeFrame) const;
+
+  /// Return number of ShapeFrames added to this CollisionGroup
+  size_t getNumShapeFrames() const;
+
+  /// Return index-th ShapeFrames registered to this CollisionGroup. Return
+  /// nullptr if the index is out of the range.
+  const dynamics::ShapeFrame* getShapeFrame(size_t index) const;
+
+  /// Return all the ShapeFrames registered to this CollisionGroup
+  const std::vector<const dynamics::ShapeFrame*> getShapeFrames() const;
+
+  /// Update engine data. This function should be called before the collision
+  /// detection is performed by the engine in most cases.
+  void update();
+
+  /// Perform collision detection within this CollisionGroup.
+  bool detect(const Option& option, Result& result);
+
+  /// Perform collision detection with other CollisionGroup.
+  ///
+  /// Return false if the engine type of the other CollisionGroup is different
+  /// from this CollisionObject engine.
+  bool detect(CollisionGroup* group, const Option& option, Result& result);
+
+  /// Return all the CollisionObjects in this CollisionGroup
+  const std::vector<CollisionObject*>& getCollisionObjects();
+
+protected:
+
+  /// Initialize the collision detection engine data such as broadphase
+  /// algorithm. This function will be called whenever ShapeFrame is either
+  /// added to or removed from this CollisionGroup.
+  virtual void initializeEngineData() = 0;
+
+  /// Notify that a CollisionObject is added so that the collision detection
+  /// engine do some relevant work
+  virtual void notifyCollisionObjectAdded(CollisionObject* object) = 0;
+
+  /// Notify that CollisionObjects are added so that the collision detection
+  /// engine do some relevant work
+  virtual void notifyCollisionObjectsAdded(
+      const std::vector<CollisionObject*>& collObjects) = 0;
+
+  /// Notify that a CollisionObject is removed so that the collision detection
+  /// engine do some relevant work
+  virtual void notifyCollisionObjectRemoved(CollisionObject* object) = 0;
+
+  /// Notify that all the CollisionObjects are remove so that the collision
+  /// detection engine do some relevant work
+  virtual void notifyAllCollisionObjectsRemoved() = 0;
+
+  /// Update the collision detection engine data such as broadphase algorithm.
+  /// This function will be called ahead of every collision checking.
+  virtual void updateEngineData() = 0;
+
+protected:
```

Comment:
Is it safe for the user to modify these? My understanding is that these fields should *only* be modified by the non-`virtual` `public` methods on this class. If so, we should make them `private`.